### DISCUSSION TITLE
Question about GINConv

### DISCUSSION THREAD

**Discussion by Bowen-n** [pyg-team/pytorch_geometric#2250] [Q&A] [answered]
```python
    def forward(self, x: Union[Tensor, OptPairTensor], edge_index: Adj,
                size: Size = None) -> Tensor:
        """"""
        if isinstance(x, Tensor):
            x: OptPairTensor = (x, x)

        # propagate_type: (x: OptPairTensor)
        out = self.propagate(edge_index, x=x, size=size)

        x_r = x[1]
        if x_r is not None:
            out += (1 + self.eps) * x_r

        return self.nn(out)
```
I don't understand the code `x: OptPairTensor = (x, x)` in `forward` of [GINConv](https://pytorch-geometric.readthedocs.io/en/latest/_modules/torch_geometric/nn/conv/gin_conv.html#GINConv). Why two `x` are needed? The first `x` is used in `message_and_aggregate` and the second is used in `out += (1 + self.eps) * x_r`

**Comment by wsad1** [comment, marked as answer]
GINConv, like some other message passing layers work on graphs with symmetric adjacency matrices and bipartite graphs. read more [here](https://pytorch-geometric.readthedocs.io/en/latest/notes/create_gnn.html). Bipartite graphs can be marked by passing x=(x_n,x_m) or a pair of tensors.

If a graph where x is a `Tensor` is passed.  This line `  x: OptPairTensor = (x, x)`  converts this graph to an equivalent bipartite graph. This way the rest of the code can be written assuming a bipartite graph is passed. But here  `x[0]=x[1]`, and you could use them interchangeably.
However if a bipartite graph is passed and x is an `OptPairTensor` then x[0] might not be equal to x[1], they might even have different lengths. In this case the edges are going from x_l(x[0]) to x_r(x[1]). So the`message_and_aggregate` will aggregate messages from x[0].  The aggregated features will be added to x_r via `out += (1 + self.eps) * x_r`.
